drivers: add imx7d gpio driver

diegosueiro opened the pull request:
This pull request is to add gpio driver for mcimx7_m4 arch.

The implementation was tested in the colibri_imx7d_m4 board by compiling and running the samples/basic/button application. All possible interrupt variations were validated.

diegosueiro commented:
In order to get all tests passed the PR #7074 is needed.

dbkinder reviewed (approved):
+1 for docs

stanislav-poboril commented on drivers/gpio/gpio_imx.c:
```diff
@@ -0,0 +1,387 @@
+		GPIO_Init(config->base, &pin_config);
+	} else {	/* GPIO_ACCESS_BY_PORT */
+		for (i = 0; i < 32; i++) {
+			pin_config.pin = pin;
```
pin_config.pin = i;

diegosueiro replied:
Good catch

stanislav-poboril commented on drivers/gpio/gpio_imx.c:
```diff
@@ -0,0 +1,387 @@
+	const struct imx_gpio_config *config = dev->config->config_info;
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		*value = GPIO_ReadPinOutput(config->base, pin);
```
Shouldn't be GPIO_ReadPinInput() used here? Anyway GPIO_ReadPinInput() and GPIO_ReadPinOutput() does the same thing in BSP driver ext/hal/nxp/imx/drivers/gpio_imx.h.

diegosueiro replied:
OK

stanislav-poboril commented on drivers/gpio/gpio_imx.c:
```diff
@@ -0,0 +1,387 @@
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		*value = GPIO_ReadPinOutput(config->base, pin);
+	} else { /* GPIO_ACCESS_BY_PORT */
+		*value = GPIO_ReadPortOutput(config->base);
```
Shouldn't be GPIO_ReadPortInput() used here? Anyway GPIO_ReadPortInput() and GPIO_ReadPortOutput() does the same thing in BSP driver ext/hal/nxp/imx/drivers/gpio_imx.h.

diegosueiro replied:
OK

stanislav-poboril commented on ext/hal/nxp/imx/drivers/gpio_imx.c:
```diff
@@ -68,28 +70,33 @@ void GPIO_Init(GPIO_Type* base, const gpio_init_config_t* initConfig)
+        }
+        case(gpioIntRisingOrFallingEdge):
+        {
+            GPIO_EDGE_SEL_REG(base) |= (1U << pin);
```
There is a GPIO_SetIntEdgeSelect() function which does this setup.
It is possible to use gpioNoIntmode and call GPIO_SetIntEdgeSelect() afterwards, in which case the modification of this external BSP driver would perhaps be unnecessary.
But adding the new enum value gpioIntRisingOrFallingEdge seems cleaner.

diegosueiro replied:
I don't think that using gpioNoIntmode reflects what we want to configure. This can be confusing for someone reading the code. I had to change this code anyway for the IMR and EDGE_SEL registers cleanup and the pin variable assignment.
I'll change it to call the GPIO_SetIntEdgeSelect() function instead.

diegosueiro commented:
@MaureenHelm,

Any feedback on this PR?

MaureenHelm commented:
Hi @diegosueiro , I'd really like to avoid modifying the imported code. Do you have any issues using @stanislav-poboril's gpio driver instead?

diegosueiro commented:
@MaureenHelm,

No problem. We can use the gpio shim driver from PR #7123.
As soon as it gets merged I will update this one to add support for imx7 and Colibri board. OK?

stanislav-poboril commented:
I have extracted GPIO driver shim from #7123 into #7306. You can review it there and rebase this PR once it is merged.

diegosueiro commented:
Thanks @stanislav-poboril .

@MaureenHelm ,
I already rebased my PR and made the appropriate adjustments and tests.

MaureenHelm commented on boards/arm/colibri_imx7d_m4/board.h:
```diff
@@ -9,4 +9,10 @@
+#define LED0_GPIO_PIN   2
+
+#define SW0_GPIO_NAME   CONFIG_GPIO_IMX_PORT_2_NAME
+#define SW0_GPIO_PIN    26
```
We can now get this information from dts. Can you remove these macros and add `leds` and `gpio_keys` nodes to the board dts? See frdm_k64f for an example.

MaureenHelm commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -122,7 +128,7 @@ The Colibri iMX7D doesn't have QSPI flash for the M4 and it needs to be started
 RAM, put the M4 in reset, set the M4 Program Counter and Stack Pointer, and get
 the M4 out of reset.
-The A7 can perform these steps at bootloader level after the Linux system has
+The A7 can perform these steps at bootloader level or after the Linux system has
```
Does this change and the one on line 192 have anything to do with gpio? If not, please make the change in a separate commit.

MaureenHelm commented on boards/arm/colibri_imx7d_m4/pinmux.c:
```diff
@@ -11,6 +11,21 @@ static int colibri_imx7d_m4_pinmux_init(struct device *dev)
+#ifdef CONFIG_GPIO_IMX_PORT_1
+	/* GPIO1_IO02 Mux Config */
+	*(&IOMUXC_LPSR_SW_MUX_CTL_PAD_GPIO1_IO02) = 0;
+	*(&IOMUXC_LPSR_SW_PAD_CTL_PAD_GPIO1_IO02) = 0;
```
Why are you getting a reference and immediately dereferencing it? This seems unnecessary

diegosueiro committed:
doc: fix colibri_imx7d_m4 board documentation

Fix a typo and update u-boot command

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@MaureenHelm ,

Thanks for the review. Changes applied.

MaureenHelm commented on boards/arm/colibri_imx7d_m4/pinmux.c:
```diff
@@ -11,6 +11,15 @@ static int colibri_imx7d_m4_pinmux_init(struct device *dev)
+	*(&IOMUXC_SW_PAD_CTL_PAD_EPDC_GDRL) =
+		IOMUXC_SW_PAD_CTL_PAD_EPDC_GDRL_PS(2) |
+		IOMUXC_SW_PAD_CTL_PAD_EPDC_GDRL_PE_MASK |
+		IOMUXC_SW_PAD_CTL_PAD_EPDC_GDRL_HYS_MASK;
```
Same question as before about getting a reference and immediately dereferencing it. Why not use `IOMUXC_SW_MUX_CTL_PAD_EPDC_GDRL` and `IOMUXC_SW_PAD_CTL_PAD_EPDC_GDR` directly?

MaureenHelm commented on arch/arm/soc/nxp_imx/mcimx7_m4/soc.c:
```diff
@@ -51,6 +51,30 @@ void SOC_RdcInit(void)
+#ifdef CONFIG_GPIO_IMX_PORT_1
+	RDC_SetPdapAccess(RDC, rdcPdapGpio1,
+		RDC_DOMAIN_PERM(CONFIG_DOMAIN_ID, RDC_DOMAIN_PERM_RW),
+		false, false);
```
I apologize for not being more specific before. Please align lines 60-61 with the first argument in line 59. Do the same below for lines 69-70.

MaureenHelm replied:
You might want to check your vim settings and set tabstop=8. Sometimes the github diff viewer does funny things with alignment, but I checked out your code locally and it looks fine.

diegosueiro committed:
board: add gpio support for colibri_imx7d_m4 board

Adds definitions, configurations and device tree entries for
colibri_imx7d_m4 board.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
arch: Add imx7d_m4 gpio definitions

Adds all necessary gpio definitions and configurations for imx7d_m4 soc.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>
